### PR TITLE
Goto-cc extensions to build (and link) recent Linux kernels

### PR DESCRIPTION
**Pull request by tautschnig**
The first two commits are nice-to-have to avoid some warnings, the support for thin archives is essential.

### CODE REVIEW

File: src/goto-cc/compile.cpp [hunk 1]
```diff
@@ -135,56 +136,113 @@ bool compilet::doit()
...
+};
+
+static file_typet detect_file_type(const std::string &file_name)
+{
+  // first of all, try to open the file
+  std::ifstream in(file_name);
+  if(!in)
+    return file_typet::NO_SUCH_FILE;
+
+  const std::string::size_type r = file_name.rfind('.', file_name.length() - 1);
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-cc/compile.cpp]: Why not just `file_name.rfind('.')`?

File: src/goto-cc/compile.cpp [hunk 2]
```diff
@@ -135,56 +136,113 @@ bool compilet::doit()
...
+  GOTO_BINARY,
+  ELF_OBJECT
+};
+
+static file_typet detect_file_type(const std::string &file_name)
+{
+  // first of all, try to open the file
+  std::ifstream in(file_name);
+  if(!in)
+    return file_typet::NO_SUCH_FILE;
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-cc/compile.cpp]: Opening a file can fail for other reasons than the file not existing, although as far as I'm aware there's no portable way to detect this in C++11 so I guess this is fine.

Comment (tautschnig) [reply src/goto-cc/compile.cpp]: I have renamed the enum value and also made the warning message more detailed.

File: src/goto-cc/compile.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-cc/compile.cpp]: Copy&paste from the previous code, I've simplified it now.

**Review by kroening — APPROVED** [review]

**Pull request merged**